app/controllers/api/base_controller.rb

```diff
@@ -37,11 +37,18 @@ def verify_mfa_requirement
     render plain: "Gem requires MFA enabled; You do not have MFA enabled yet.", status: :forbidden
   end
 
+  def verify_api_key_gem_scope
+    return unless @api_key.rubygem && @api_key.rubygem != @rubygem
+
+    render plain: "You do not have permission. This API key is scoped to #{@api_key.rubygem.name}.", status: :forbidden
```

Comment:
Should we really report back which gem is assigned to the key? I would expect same message with wrong (non-existing key) as for key with different scope. It is just invalid for this request (and it doesn't matter why). Exposing this info can make it it easier when you find out a key without having knowledge which gem is assigned to it.

Yes, in theory you can try to find all possible gems to the key and try one by one, but what to make it easier?

Reply:
My intent was to display why the request was invalid as much as possible, but it's true that we are making it easier for someone malicious to use this key as you mentioned. I think it's sufficient to say that the key isn't scoped to this gem, and the user can view the key on the UI to figure out which gem the key belongs to 👍

Reply:
I think "invalid key" message and "key belongs to different gem" messages should be equal.

Reply:
I see what you mean. However, the behaviour between these two cases has a slight difference. For the invalid case, the key cannot be used in **any** api request while the key belongs to a different gem cannot be used in some **owner and gem specific commands**, so I'm having a bit of trouble crafting a consistent message. 

For now, I have it as `"Rubygem <name> cannot be scoped this API key".` but that does not sounds entirely correct for either cases. Instead I was thinking of changing  in the "key belongs to a different gem" message to be something like this:

`This API key does not have permission to modify this gem`

while in the "invalid key case" have something more generic since it is set up so that there could be multiple reasons why something is invalid (eg. implementing the concept of expiries on keys).

`This API key is invalid`

wdyt?